### PR TITLE
Local twilio verification

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
This PR adds support for generating local verification codes instead of having to make a HTTP request to Twilio to do so. This results in a faster verification process with the tradeoff of a higher potential of failure due to the fact that the user is now in charge of managing the verification.

Of course, we still need to inform Twilio that the verification was completed, so we introduce support for that in a queued listener that is dispatched after a verification occurs. This is implemented in such a way as to allow for additional services that work in the same way to plug into this lifecycle by implementing the `InformsServiceAfterVerification` contract.

### CODE REVIEW

File: src/Services/Twilio/TwilioVerificationService.php [hunk 1]
```diff
@@ -18,25 +23,26 @@
...
-    public function __construct(Client $twilio, string $verifyId)
-    {
-        $twilio = $twilio->verify->v2->services($verifyId);
-        $this->verifications = $twilio->verifications;
-        $this->verificationChecks = $twilio->verificationChecks;
+    public function __construct(
+        Client $twilio,
+        string $verifyId,
+        private ?VerificationCodeManager $verificationCodeManager,
+        private ?Dispatcher $event,
```
Comment (olivernybroe) [inline src/Services/Twilio/TwilioVerificationService.php:37]: I think we should always dispatch the verified event, so let's just make this a requirement 🤔

File: src/VerificationCodeManagers/CacheVerificationCodeManager.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public function __construct(
+        private Repository $cache,
+        private VerificationCodeGenerator $codeGenerator,
+    ) {
+    }
+
+    public function store(PhoneNumber $phoneNumber): string
+    {
+        $code = $this->codeGenerator->generate($phoneNumber);
```
Comment (olivernybroe) [inline src/VerificationCodeManagers/CacheVerificationCodeManager.php:22]: I feel like the verification code manager should not call the generator 🤔 

What if we call it a `VerificationCodeStore`, also to make it clearer that the purpose of it is to store a code and fetch it.

**Review by olivernybroe — COMMENTED** [review]
I think the idea is great, however there is something here that seems a bit wrong. 
Not entirely sure what, but something seems a bit too complex or like it is tied to Twilio without being totally tied.

I think what might be the problem for me is the `generate_code_locally` thing, it doesn't seem that intuitive... 

I think maybe a bigger restructure or thought process here is needed 🤔 


Let's try and build up what we are doing
- Generate a code (`CodeGenerator`) 
- Send the code (`CodeSender`) 
- Verify the code (`CodeVerifier`) 

By default when using Twilio, it will generate the code for us, meaning the code generator should be `NullCodeGenerator`  
The `TwilioCodeSender` will then receive a `Code` object and phone number, if the code object is not null, then it will use the custom code option when sending it.

When verifying, the `CodeVerifier` receives a `Code` object and phone number and send a verify call to the API to check if it is valid.


In case `'cache' => true`, then we check if the current driver implements `Cacheable`, it not throw exception. 
`CodeGenerator` should now be replaced with `NumericCodeGenerator` and both `CodeSender` and `CodeVerifier` should now be extended with `CachedCodeSender` and `CachedCodeVerifier`. 

Now the `CachedCodeSender` can use the `CodeSender` to send like normally and cache it afterwards, then the `CachedCodeVerifier` can verify the code from cache and queue up the normal `CodeVerifier` 🤔 

What do you think of this? Is it more confusing or not? 
We should prob. take a quick call 😆

**Comment by lukeraymonddowning** [comment]
@olivernybroe let's book some time in after holidays and Tuple on this one 👍

**Comment by lukeraymonddowning** [comment]
Going to close this for now as a lot of water has gone under the bridge and we don't really have requirements for it.

**Pull request closed without merging**